### ISSUE TITLE
Use Infill to transition Filament?

### ISSUE THREAD

**Issue by cfelicio** [Ultimaker/Cura#3657]
Pretty cool idea, not sure how hard it would be to implement:

https://mosaicmanufacturing.zendesk.com/hc/en-us/articles/115002756094-Minimizing-Filament-Waste

Basically, instead of printing a large, solid prime tower when switching filaments, this attempts to purge the filament inside the part (as infill). This means:

1 - Infill needs to be printed first
2 - The object needs to be large enough to provide enough infill
3 - The prime tower need to offer variable infill (100% infill at top / bottom layers, sparse prime tower when printing infill)

Considering dual / multi color are becoming more prevalent, and Cura seems to be implementing a lot of exciting features recently (variable layer height, tree support, etc), wondering if the devs would be willing to take a look at this one? :-)

**Comment by BagelOrb** [comment]
To me tis feature doesn't make sense. If you want to minimize material then you should only print material where needed.

So you can only reduce material where it's not needed.

The infill material *is* needed, however. 

Rather we should minimize the material for infill by looking at where we don't need as much material, rather than doing it in the uncontrolled fashion you propose.

I think this issue should be a [Won't fix].

**Comment by ianpaschal** [comment]
> If you want to minimize material then you should only print material where needed.
> The infill material is needed, however.

But that's kind of his point right? Rather than "waste" material printing a prime tower, use the infill to prime?

**Comment by BagelOrb** [comment]
If you don't need a prime tower than don't print it!

The whole idea of the prime tower is to make sure we are extruding correctly. If we skip that and print infill instead we know we are printing infill incorrectly, while infill serves a purpose and so it should be printed correctly.

**Comment by cfelicio** [comment]
Im not proposing to skip the prime tower, just have a dynamic one. Example:

Layer 1 - 4: Prime tower 100% infill (as there is no infill printed yet)
Layer 5 - 20: Prime tower 20% infill (rest goes inside the part as infill)
Layer 20 - 25: Prime tower 30% infill

I understand that the feature might not sound that useful for old style setups, where you have multiple nozzles, but the new gen of printers are all trying to go for the multiple extruders with a single nozzle, and that requires a much larger prime tower, wasting a lot more filament that could be used to infill instead.

The approach mosaic took (as shown on the link) is quite cool, but proprietary. Cura seems to be implementing quite a few innovative features, hence why I thought I'd ask! :-)

**Comment by BagelOrb** [comment]
I don't agree in your division of printers into old and new style. The Y-splitter has been around for a long time - also for Ultimaker printers. It's a cheap solution for hardware, but causes a rediculous amount of material waste and also a waste of space for the build plate. But that's just semantics ;)

The prime tower in Cura was never meant for Y-splitters and I don't think Ultimaker will spend time on implementing a feature for Y-splitter printers. However we will consider pull requests if somebody writes code for Y-splitter printers.

**Comment by Ghostkeeper** [comment]
This ties in with https://github.com/Ultimaker/Cura/issues/3550 and https://github.com/Ultimaker/Cura/issues/3553 to make it possible to have a lower-density prime tower for one of the extruders.

**Comment by BagelOrb** [comment]
No, the proposal here is to just print infill instead of printing a prime tower and purging.

**Comment by cfelicio** [comment]
Hi Tim,

Sorry if I wasn't clear on the request.  I don't think the proposal is to only print infill without the tower, that's not possible as for some areas of the print there is not enough infill to do the transition, so a tower is still required.

The idea is to actually to have a tower with dynamic fill. Example:

Layer 1 - 4: Prime tower 100% infill (as there is no infill printed yet)
Layer 5 - 20: Prime tower 20% infill (rest goes inside the part as infill)
Layer 20 - 25: Prime tower 30% infill

Thanks! :-)

**Comment by cfelicio** [comment]
This basically takes #3550 one step further.

**Comment by sjoerdtimmer** [comment]
I like the idea but I foresee the following problem: When you switch from 20% to 30% infill most infill lines will not align with the previous layer. If you do that multiple times all strength will be gone from the structure. Normal infill works because the infill line-line-distance does not change (or only a little with some of the fancy new infill patterns). There are two possible solutions: 1) only halve or double the line-line distance between layers. That way they will align maximally where possible. 2) let infill lines gradually move towards each other and merging when not needed and split (in advance) when they are needed again. This would require a completely new infill type...

**Comment by Ghostkeeper** [comment]
In 3.5 with the new prime tower strategy it would be feasible to prime less for one of the extruders on some layers. We don't require any more that the inner prime towers are resting on a previous layer since they attach to the side of the tower. We do require that the outer prime tower is sturdy so one of the extruders won't be able to do this.

**Comment by sjoerdtimmer** [comment]
I was referring to the proposal to change from 20% infill to 30% infill between layers. That would give problems I believe.

**Comment by Nerdy314159265** [comment]
I would thing something like Rectilinear would already work since it has that alternating pattern to keep a relatively narrow gap and there are no paths that the higher % layers would take that wouldn't have some support under them.

**Comment by cptSwing** [comment]
I'd love to see this feature in Cura! Just leaving a link to PrusaSlicer here, I haven't used it myself --> https://help.prusa3d.com/en/article/wipe-tower_125010 ("wipe to infill")

**Comment by m-sundman** [comment]
This would be an exceptionally awesome feature!
A few comments:
1. Also print infill after changing the filament, if at all possible, especially if there's no other priming possibility (skirts and brims usually can't be used).
2. Increasing the infill percentage of the prime tower is trivial. It doesn't need an even infill, so just use, say, lines, and then lines can be added/omitted on a layer-by-layer basis without any problem. (E.g., it's perfectly fine if a part of the layer in the prime tower has lines every 2 mm and a part of the layer has lines every 1 mm.)
3. It would be great to be able to have the "purge/prime" infill have higher density and/or flow rate. Having some layers have more infill than others is usually perfectly fine, and having some extra infill could be particularly useful when the purging/priming is causing uneven infill flow.

**Comment by wblake0000** [comment]
The "Wipe to Infill" feature is the most logical implementation as opposed to a prime tower.  The infill density doesn't have to change , simply use the infill to wipe the nozzle instead of a prime tower.  It's a very efficient use of travel moves and filament.  I would love to see this feature added, and I love Cura BTW.  There are obvious drawbacks with dark filament infill in light color walls and narrow points of the print without much infill where the wipe could be insufficient but I think for a lot of prints it might be an excellent option.

**Comment by Juusoz** [comment]
I don't see why not, you just need to swap the extruder, print an infill line, count the E-steps into some sort of _"InfillEStepsOnLayer"_ variable, then keep repeating it until you've done all of the infill on the layer.

Then you just need to subtract that amount of volume out of the prime tower, although that might be a bit difficult with the way the prime tower is currently implemented in Cura. The easiest way to fix it is probably to just change the prime tower into a regular cylindrical model that prints with infill, then put the default infill at 100% and lower it as necessary. Also, "_Lines_" infill would probably work the best in this situation.

**Comment by Nerdy314159265** [comment]
It could always just have a minimum volume or minimum infill% on the prime tower that it snaps to in order to prevent the prime tower skipping full layers and falling apart.

**Comment by cfelicio** [comment]
Oh wow, I didn't realize I had this one still open, from 2018! Time flies!

In any event, here is what happened:

1 - I ditched the Y splitter, it was not very reliable
2 - I bought a Palette, but their slicer is terrible
3 - Someone created a fantastic script called p2pp that supports all these fancy features, but sadly only on the prusa slicer, so I had to switch from cura to prusa

Here is the reference documentation for the wipe to infill, perhaps some of this code can be backported to cura at some point, if there is interest: https://help.prusa3d.com/article/wipe-tower_125010

I have been using it flawlessly and it helps reduce the size of the wipe tower (but not completely eliminate it).

**Comment by Hello1024** [comment]
Something not mentioned yet:

Typically the 'prime tower' consists of both priming the nozzle (getting rid of air bubbles), but also purging the previous color of filament.

Air bubbles in the infill is probably acceptable...    As is the 'wrong color'.

However, as noted above, in some cases it might not be possible to fully purge the nozzle within the infill - particularly on the top and bottom few layers.


Therefore, the logic should probably go as follows:

  * Change filament
  * Print infill first (already the default).
  * Automatically adjust infill density as necessary to use up the purge_distance of filament in any non-external layer.
  * If infill density hits 100% on a given layer, then put the remainder in a purge tower.
  * The purge tower must be built anyhow (unless the printer has a wiper).

**Comment by m-sundman** [comment]
Yes, exactly like Hello1024 says, with the additional detail that printing into the purge tower should be done first, before going to the infill, just to make sure the worst part goes into the tower instead of into the infill.
